☣️[stdlib] Revisit underestimatedCount default implementation for Collection

moiseev opened the pull request:
Currently the default implementation of `underestimatedCount` property
for a `Collection` is to simply return `count`. Problem with that is
`count` is allowed to be O(n) operation on some collections and can have
a side-effect of iterating over the whole collection, which defeats the
whole purpose of `underestimatedCount`.

The proposed change makes `underestimatedCount` strictly O(1), which
requires a certain pessimization of it's return value for `Collection`
instances that are only forward or bidirectional.
`RandomAccessCollection` instances guarantee constant time `count` and
therefore still can provide a default implementation of
`underestimatedCount` in terms of `count`.

**UPD**: Accidentally addresses https://bugs.swift.org/browse/SR-6693

Commit:
[stdlib] Revisit underestimatedCount default implementation for Collection

Currently the default implementation of `underestimatedCount` property
for a `Collection` is to simply return `count`. Problem with that is
`count` is allowed to be O(n) operation on some collections and can have
a side-effect of iterating over the whole collection, which defeats the
whole purpose of `underestimatedCount`.

The proposed change makes `underestimatedCount` strictly O(1), which
requires a certain pessimization of it's return value for `Collection`
instances that are only forward or bidirectional.
`RandomAccessCollection` instances guarantee constant time `count` and
therefore still can provide a default implementation of
`underestimatedCount` in terms of `count`.

moiseev commented:
@swift-ci Please benchmark

moiseev commented:
@swift-ci Please Test Source Compatibility

milseman commented:
Hooray, an API I can actually use

CTMacUser commented:
Instead of a fixed `0`, a "`isEmpty ? 0 : 1`" would be slightly more useful, and still O(1).

xwu commented:
Could isEmpty have side effects? If so, then it’d be a pretty hard thing to debug.

moiseev commented:
@swift-ci Please benchmark
...because I can't believe in these numbers.

moiseev commented:
@swift-ci Please smoke test

milseman commented:
@moiseev the Queue benchmark regressions are likely real, as they are generic over RRC?

milseman commented:
@swift-ci please smoke test

Commit:
Implement ReversedCollection.underestimatedCount

ReversedCollection can delegate to its base's underestimatedCount
instead of using the default implementation that, since
ReversedCollection is only bidirectional, would return 0.

Commit:
Remove redundant implementation of popLast from the benchmark

moiseev commented:
@swift-ci Please benchmark

moiseev commented:
@milseman The `Queue` benchmark regressions were related to now-unoptimal `ReversedCollection.underestimatedCount`, that, being only a `BidirectionalCollection` would now return `0` instead of `count`. Fixed now.

Commit:
Implement <IntegerType>.Words.underestimatedCount

moiseev commented:
@swift-ci Please benchmark

milseman reviewed (approved):
LGTM.

BTW, what about all the lazy? E.g. LazyMapSequence should also forward to its wrapped Sequence. Do we have other Sequence/Collections that are just wrappers/adaptors?

moiseev commented:
Interesting... Sim2DArray does not show this regression when I run it locally...

moiseev commented:
@swift-ci Please test

milseman commented:
Sim2DArray is an extremely noisy benchmark

Commit:
Fix the underestimatedCount default impl test

moiseev commented:
@swift-ci Please smoke test

Commit:
Implement underestimatedCount for hashed collections

moiseev commented:
@swift-ci Please smoke test

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
@swift-ci Please smoke test

Commit:
Fix the Lazy validation-test after the reversed.underestimatedCount change

moiseev commented:
@swift-ci Please smoke test

benrimmington commented on stdlib/public/core/Dictionary.swift:
```diff
@@ -584,6 +584,15 @@ extension Dictionary: Sequence {
     return _variantBuffer.makeIterator()
   }
+
+  /// A value less than or eequal to the number of key-value pairs in the
```
Typo: `eequal`

benrimmington commented on stdlib/public/core/Set.swift:
```diff
@@ -245,6 +245,14 @@ extension Set: Sequence {
     return _variantBuffer.makeIterator()
   }
 
+  /// A value less than or eequal to the number of elements in the set.
```
Typo: `eequal`
